File: gru.py
import time
end = time.time()

import torch
import torch.nn as nn
from torch.autograd import Variable

class Uniskip(nn.Module):

    def __init__(self, input_size, hidden_size):
        super(Uniskip, self).__init__()
        self.gru = nn.GRU(input_size=input_size,
                          hidden_size=hidden_size,
                          batch_first=True)
 
    def _select_last(self, input, lengths):
        x = Variable(input.data.new().resize_((input.size(0), input.size(2))))
        for i in range(input.size(0)):
            x[i] = input[i,lengths[i]-1]
        return x
 
    def forward(self, input, lengths=[26]*100):
        x, hn = self.gru(input)
        x = self._select_last(x, lengths)
        return x

nbatch = 100
seq_len = 26
input_size = 620
hidden_size = 2400
batch_size = 100

model = Uniskip(input_size, hidden_size)
model = model.cuda()
criterion = nn.MSELoss().cuda()

input = Variable(torch.randn(batch_size, seq_len, input_size)).cuda()
target = Variable(torch.randn(batch_size, hidden_size)).cuda()

loss = criterion.forward(model.forward(input), target)
loss.backward()
torch.cuda.synchronize()
print("Setup : compile + forward/backward x 1")
print("--- {} seconds ---".format(time.time() - end))


end = time.time()
for i in range(nbatch):
   model.forward(input)
torch.cuda.synchronize()
print("Forward:")
nSamples = nbatch * batch_size
speed = nSamples / (time.time() - end)
print("--- {} samples in {} seconds ({} samples/s, {} microsec/samples) ---".format(
        nSamples, (time.time() - end), speed, 1000000/speed))


end = time.time()
for i in range(nbatch):
    loss = criterion.forward(model.forward(input), target)
    loss.backward()
torch.cuda.synchronize()
print("Forward + Backward:")
speed = nSamples / (time.time() - end)
print("--- {} samples in {} seconds ({} samples/s, {} microsec/samples) ---".format(
        nSamples, (time.time() - end), speed, 1000000/speed))

File: gru.lua
local timer = torch.Timer()

require 'nn'
require 'rnn'

require 'cutorch'
require 'cunn'

local nbatch = 100
local seqlen = 26
local batchsize = 100 --256
local inputsize = 620
local hiddensize = 2400

function GRU(inputsize, hiddensize, trimzero)
   local gru = nn.GRU(inputsize, hiddensize, false, 0, true)
   if trimzero then
      gru:trimZero(1)
   end
   local model = nn.Sequential()
      :add(nn.SplitTable(2))
      :add(nn.Sequencer(gru))
      :add(nn.SelectTable(-1))
   return model
end

local model = GRU(inputsize, hiddensize, true):cuda()

local input = torch.rand(batchsize, seqlen, inputsize):cuda()
local target = torch.rand(batchsize, hiddensize):cuda()
input[{{},{1,20},{}}]:fill(0)

local criterion = nn.MSECriterion():cuda()

criterion:forward(model:forward(input), target)
model:backward(input, criterion:backward(model.output, target))
cutorch.synchronize()
print("Setup : compile + forward/backward x 1")
print("--- " .. timer:time().real .. " seconds ---")

timer:reset()
for i=1, nbatch do
   model:forward(input)
end
cutorch.synchronize()
print("Forward:")
local nSamples = nbatch * batchsize
local speed = nSamples / timer:time().real
print("--- " .. nSamples .. " samples in " .. timer:time().real.. " seconds (" .. speed .. " samples/s, " .. 1000000/speed .. " microsec/samples) ---")

timer:reset()
for i = 1, nbatch do
   criterion:forward(model:forward(input), target)
   model:zeroGradParameters()
   model:backward(input, criterion:backward(model.output, target))
   model:updateParameters(0.01)
end
cutorch.synchronize()
print("Forward + Backward:")
local speed = nSamples / timer:time().real
print("--- " .. nSamples .. " samples in " .. timer:time().real .. " seconds (" .. speed .. " samples/s, " .. 1000000/speed .. " microsec/samples) ---")

